## Load the dataset

In [ ]:
# Import libraries
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
from sklearn.model_selection import train_test_split

In [ ]:
# Download the dataset
!wget https://www.webpages.uidaho.edu/vakanski/Codes_Data/spam_messages.csv

--2025-03-28 19:28:32--  https://www.webpages.uidaho.edu/vakanski/Codes_Data/spam_messages.csv
Resolving www.webpages.uidaho.edu (www.webpages.uidaho.edu)... 129.101.105.230
Connecting to www.webpages.uidaho.edu (www.webpages.uidaho.edu)|129.101.105.230|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 45108305 (43M) [application/octet-stream]
Saving to: ‘spam_messages.csv’

spam_messages.csv   100%[===================>]  43.02M  11.0MB/s    in 4.3s    

2025-03-28 19:28:37 (10.0 MB/s) - ‘spam_messages.csv’ saved [45108305/45108305]



In [ ]:
# Load the CSV file
dataset = pd.read_csv('spam_messages.csv', index_col=False, encoding = 'latin-1')

# Display the shape of the dataset
print('Data shape', dataset.shape)

# Display the first 10 messages in the dataset
dataset.head(10)

Data shape (47392, 2)


label                                               text
0   ham  Funny fact Nobody teaches volcanoes 2 erupt, t...
1   ham  I sent my scores to sophas and i had to do sec...
2  spam  We know someone who you know that fancies you....
3   ham  Only if you promise your getting out as SOON a...
4  spam  Congratulations ur awarded either �500 of CD...
5   ham         I'll text carlos and let you know, hang on
6   ham          K.i did't see you.:)k:)where are you now?
7   ham             No message..no responce..what happend?
8   ham  Get down in gandhipuram and walk to cross cut ...
9   ham                         You flippin your shit yet?

In [ ]:
# Replace 0 with 'ham' and 1 with 'spam' in the labels
dataset['label'] = dataset['label'].replace({'ham':0, 'spam':1})

dataset

<ipython-input-4-b094fd578c8d>:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dataset['label'] = dataset['label'].replace({'ham':0, 'spam':1})


label                                               text
0          0  Funny fact Nobody teaches volcanoes 2 erupt, t...
1          0  I sent my scores to sophas and i had to do sec...
2          1  We know someone who you know that fancies you....
3          0  Only if you promise your getting out as SOON a...
4          1  Congratulations ur awarded either �500 of CD...
...      ...                                                ...
47387      0  please change the sitara tickets dating back t...
47388      1  from mr . silas sankoh ecobank benin rue du go...
47389      0  just to get things stirred up : the estate wil...
47390      0  sorry for the oversight . bill - - - - - origi...
47391      1  executive hospitality corp ( ehpc ) operator o...

[47392 rows x 2 columns]

In [ ]:
# Define a list with class labels
class_names = ['ham', 'spam']

In [ ]:
# Display fully the first few messages (can you guess which one is spam without looking at the label?)
print(dataset.iloc[0,1])
print(dataset.iloc[1,1])
print(dataset.iloc[2,1])
print(dataset.iloc[3,1])
print(dataset.iloc[5,1])

Funny fact Nobody teaches volcanoes 2 erupt, tsunamis 2 arise, hurricanes 2 sway aroundn no 1 teaches hw 2 choose a wife Natural disasters just happens
I sent my scores to sophas and i had to do secondary application for a few schools. I think if you are thinking of applying, do a research on cost also. Contact joke ogunrinde, her school is one me the less expensive ones
We know someone who you know that fancies you. Call 09058097218 to find out who. POBox 6, LS15HB 150p
Only if you promise your getting out as SOON as you can. And you'll text me in the morning to let me know you made it in ok.
I'll text carlos and let you know, hang on


In [ ]:
# Install the datasets library from Hugging Face
!pip install -q datasets fsspec

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.2/491.2 kB 24.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 183.9/183.9 kB 14.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 9.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 11.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 16.6 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torch 2.6.0+cu124 requires nvidia-cublas-cu12==12.4.5.8; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cublas-cu12 12.5.3.2 which is incompatible.
torch 2.6.0+cu124 requires nvidia-cuda-cupti-cu12==12.4.127; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cuda-cupti-cu12 12.5.82 which is incompatible.
torch 2.6.0+cu124 requires nvidia-cuda-nvrtc-cu12=

In [ ]:
from datasets import Dataset

# Split the dataset into train, test, and validation sets
train_data, test_data = train_test_split(dataset, test_size=0.2, random_state=42)
train_data, val_data = train_test_split(train_data, test_size=0.1, random_state=42)

# create the datasets from pandas dataframes
train_dataset = Dataset.from_pandas(train_data)
val_dataset = Dataset.from_pandas(val_data)
test_dataset = Dataset.from_pandas(test_data)

# remove the index column to retain only label and text columns
train_dataset = train_dataset.remove_columns(["__index_level_0__"])
val_dataset  = val_dataset .remove_columns(["__index_level_0__"])
test_dataset = test_dataset.remove_columns(["__index_level_0__"])

In [ ]:
# Check the train dataset
train_dataset

Dataset({
    features: ['label', 'text'],
    num_rows: 34121
})

## Prepare the data for model fitting

In [ ]:
from transformers import AutoTokenizer

# Tokenizers in Hugging Face are functions that convert input text into numerical format

# Load the tokenizer for the model Mobilebert
tokenizer = AutoTokenizer.from_pretrained("google/mobilebert-uncased")

config.json:   0%|          | 0.00/847 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [ ]:
# Create a function to use the loaded tokenizer and convert the text into numercial format
def tokenize(rows):
  return tokenizer(rows['text'], padding="max_length", truncation=True, max_length=128)

In [ ]:
# Tokenize the training, validation, and test datasets
tokenized_train_dataset = train_dataset.map(tokenize, batched=True)
tokenized_val_dataset = val_dataset.map(tokenize, batched=True)
tokenized_test_dataset =  test_dataset.map(tokenize, batched=True)

Map:   0%|          | 0/34121 [00:00<?, ? examples/s]

Map:   0%|          | 0/3792 [00:00<?, ? examples/s]

Map:   0%|          | 0/9479 [00:00<?, ? examples/s]

## Import a pretrained model

In [ ]:
from transformers import AutoModelForSequenceClassification

# Import the Distilbert model and specify the number of classes
model = AutoModelForSequenceClassification.from_pretrained("google/mobilebert-uncased", num_labels=2)

pytorch_model.bin:   0%|          | 0.00/147M [00:00<?, ?B/s]

Some weights of MobileBertForSequenceClassification were not initialized from the model checkpoint at google/mobilebert-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


## Train the model

In [ ]:
from transformers import TrainingArguments

# Define training arguments for the model
training_args = TrainingArguments(
    # Directory to save model checkpoints and logs
    output_dir="distilbert-emotion",
    # Evaluation strategy - evaluate the model at the end of each epoch
    eval_strategy="epoch",
    # Learning rate for the optimizer
    learning_rate=2e-5,
    # Batch size for training
    per_device_train_batch_size=64,
    # Batch size for evaluation
    per_device_eval_batch_size=64,
    # Number of training epochs
    num_train_epochs=2,
    # Weight decay to apply for regularization
    weight_decay=0.01,
    # Disable reporting to external tools (e.g., WandB, TensorBoard)
    report_to="none"
)

In [ ]:
from sklearn.metrics import accuracy_score

# Function to compute metrics during training
def compute_metrics(eval_pred):
    outputs, labels = eval_pred
    predictions = outputs.argmax(axis=-1)
    acc = accuracy_score(labels, predictions)
    return {"accuracy": acc}

In [ ]:
from transformers import Trainer

# Initialize the Trainer with the required parameters
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train_dataset,
    eval_dataset=tokenized_val_dataset,
    processing_class=tokenizer,
    compute_metrics=compute_metrics
)

model.safetensors:   0%|          | 0.00/147M [00:00<?, ?B/s]

In [ ]:
# Fit the model (about 4 minutes)
trainer.train()

TrainOutput(global_step=1068, training_loss=55246.11951841174, metrics={'train_runtime': 578.0171, 'train_samples_per_second': 118.062, 'train_steps_per_second': 1.848, 'total_flos': 1069838721838080.0, 'train_loss': 55246.11951841174, 'epoch': 2.0})

## Evaluate the Model

In [ ]:
# Evaluate the model on the test dataset
test_metrics = trainer.evaluate(eval_dataset=tokenized_test_dataset)

# Print the accuracy
print(test_metrics)

{'eval_loss': 0.0768219381570816, 'eval_accuracy': 0.9736259099061082, 'eval_runtime': 27.1909, 'eval_samples_per_second': 348.61, 'eval_steps_per_second': 5.48, 'epoch': 2.0}


## Predict the class of several test

In [ ]:
# List of sentences to evaluate
test_texts = ["You won an award! It is a free cruise trip! Call us immediately to claim it.", "Free apple phones on us for our customers, call immediately!", "I read a great book yesterday.", "March Madness has been really fun to watch this year."]

# Tokenize the sentences, pad them to the same length for batch processing
inputs = tokenizer(test_texts, return_tensors="pt", padding=True, truncation=True).to(device=0)
inputs

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


{'input_ids': tensor([[  101,  2017,  2180,  2019,  2400,   999,  2009,  2003,  1037,  2489,
          8592,  4440,   999,  2655,  2149,  3202,  2000,  4366,  2009,  1012,
           102],
        [  101,  2489,  6207, 11640,  2006,  2149,  2005,  2256,  6304,  1010,
          2655,  3202,   999,   102,     0,     0,     0,     0,     0,     0,
             0],
        [  101,  1045,  3191,  1037,  2307,  2338,  7483,  1012,   102,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0],
        [  101,  2233, 12013,  2038,  2042,  2428,  4569,  2000,  3422,  2023,
          2095,  1012,   102,     0,     0,     0,     0,     0,     0,     0,
             0]], device='cuda:0'), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 

In [ ]:
# Forward pass through the model
outputs = model(**inputs)

# Get predicted labels for each sentence
predicted_labels = outputs.logits.argmax(axis=-1).cpu().tolist()
print("Predicted Labels:", predicted_labels)

Predicted Labels: [1, 1, 0, 0]


In [ ]:
# Print results
for text, label in zip(test_texts, predicted_labels):
    print(f"Text: '{text}'; Predicted Label: {class_names[label]}")

Text: 'You won an award! It is a free cruise trip! Call us immediately to claim it.'; Predicted Label: spam
Text: 'Free apple phones on us for our customers, call immediately!'; Predicted Label: spam
Text: 'I read a great book yesterday.'; Predicted Label: ham
Text: 'March Madness has been really fun to watch this year.'; Predicted Label: ham


## Model Report

Test Accuracy: 97.36%

Predicted Class test results: Spam, Spam, Ham, Ham



## Generate adversarial examples

In [ ]:
# Install the TextAttack library
!pip install -qq textattack

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.5/54.5 kB 5.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 26.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 60.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 445.7/445.7 kB 34.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 97.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
# Download the required resource for the NLTK (Natural Language Toolkit) library
import nltk
nltk.download('averaged_perceptron_tagger_eng')

[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger_eng.zip.


True

In [ ]:
# Wrap the trained model and tokenizer for Distilbert into a textattack model
from textattack.models.wrappers import HuggingFaceModelWrapper

wrapped_model = HuggingFaceModelWrapper(model, tokenizer)

textattack: Updating TextAttack package dependencies.
textattack: Downloading NLTK required packages.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package omw to /root/nltk_data...
[nltk_data] Downloading package universal_tagset to /root/nltk_data...
[nltk_data]   Unzipping taggers/universal_tagset.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


## Apply TextFooler attack

In [ ]:
# TextFooler attack: replaces words with synonyms to fool the model
from textattack.attack_recipes import TextFoolerJin2019

attack_1 = TextFoolerJin2019.build(wrapped_model)

textattack: Downloading https://textattack.s3.amazonaws.com/word_embeddings/paragramcf.
100%|██████████| 481M/481M [01:05<00:00, 7.30MB/s]
textattack: Unzipping file /root/.cache/textattack/tmp9yqiam29.zip to /root/.cache/textattack/word_embeddings/paragramcf.
textattack: Successfully saved word_embeddings/paragramcf to cache.
textattack: Unknown if model of class <class 'transformers.models.mobilebert.modeling_mobilebert.MobileBertForSequenceClassification'> compatible with goal function <class 'textattack.goal_functions.classification.untargeted_classification.UntargetedClassification'>.


In [ ]:
from textattack.datasets import Dataset
from textattack import Attacker

# Example text sentence
example_text = " Free apple phones on us for our customers, call us immediately!"
# Correct label for the text sentence: class label 1 is 'spam'
example_label = 1

# Create a new dataset with the example text sentence and label
# The dataset is simply a list of text and label pairs
example_dataset = Dataset([[example_text, example_label]])

In [ ]:
# Apply the attack
attacker = Attacker(attack_1, example_dataset)
attacker.attack_dataset()

textattack: Attempting to attack 10 samples when only 1 are available.


Attack(
  (search_method): GreedyWordSwapWIR(
    (wir_method):  delete
  )
  (goal_function):  UntargetedClassification
  (transformation):  WordSwapEmbedding(
    (max_candidates):  50
    (embedding):  WordEmbedding
  )
  (constraints): 
    (0): WordEmbeddingDistance(
        (embedding):  WordEmbedding
        (min_cos_sim):  0.5
        (cased):  False
        (include_unknown_words):  True
        (compare_against_original):  True
      )
    (1): PartOfSpeech(
        (tagger_type):  nltk
        (tagset):  universal
        (allow_verb_noun_swap):  True
        (compare_against_original):  True
      )
    (2): UniversalSentenceEncoder(
        (metric):  angular
        (threshold):  0.840845057
        (window_size):  15
        (skip_text_shorter_than_window):  True
        (compare_against_original):  False
      )
    (3): RepeatModification
    (4): StopwordModification
    (5): InputColumnModification(
        (matching_column_labels):  ['premise', 'hypothesis']
       

 10%|█         | 1/10 [00:56<08:25, 56.14s/it]

--------------------------------------------- Result 1 ---------------------------------------------


[Succeeded / Failed / Skipped / Total] 1 / 0 / 0 / 1:  10%|█         | 1/10 [00:56<08:31, 56.79s/it]

[[1 (99%)]] --> [[0 (65%)]]

 Free [[apple]] phones on us for our [[customers]], call [[us]] [[immediately]]!

 Free [[cobbler]] phones on us for our [[consuming]], call [[we]] [[shortly]]!



+-------------------------------+--------+
| Attack Results                |        |
+-------------------------------+--------+
| Number of successful attacks: | 1      |
| Number of failed attacks:     | 0      |
| Number of skipped attacks:    | 0      |
| Original accuracy:            | 100.0% |
| Accuracy under attack:        | 0.0%   |
| Attack success rate:          | 100.0% |
| Average perturbed word %:     | 36.36% |
| Average num. words per input: | 11.0   |
| Avg num queries:              | 80.0   |
+-------------------------------+--------+


In [ ]:
# Select the first 10 samples from the test dataset to create adverasarial exampples
text_samples = test_dataset["text"][:10]
label_samples = test_dataset["label"][:10]

In [ ]:
# check the data
text_samples

['im working technical support voice process',
 'dirty pictureblyk on aircel thanks you for being a valued member heres an exclusive animated trailer for the dirty picture the insider scoop saysvidya balans sexy avatar itsy bitsy clothes cleavageshowing blouses and raunchy dance numbers in the dirty picture are receiving jawdropping reactions fans and trade pundits say the masala movie is expected to beat the musical treat rockstar the funfilled desi boyz and the action thriller don 2 dont miss it!',
 'has your mortgage search got you down are you frustrated and confused with all the different terms and quotes don t know who is telling you the truth we can solve all your problems visit our site today and in two minutes you can have us searching thousands of programs and lenders for you get the truth get the facts get your options all in one shot it s absolutely free and you can be done in only two minutes so hyperlink click right now and put your worries behind you ryte 1635465 po16354

In [ ]:
# check the true labels
label_samples

[0, 1, 1, 1, 1, 0, 0, 1, 1, 0]

In [ ]:
# Recall the class names
class_names

['ham', 'spam']

In [ ]:
# Convert the data to the required format for TextAttack
testdataset10samples_dataset = Dataset(list(zip(text_samples, label_samples)))

# Run the attack
attacker = Attacker(attack_1, testdataset10samples_dataset)
attacker.attack_dataset()

Attack(
  (search_method): GreedyWordSwapWIR(
    (wir_method):  delete
  )
  (goal_function):  UntargetedClassification
  (transformation):  WordSwapEmbedding(
    (max_candidates):  50
    (embedding):  WordEmbedding
  )
  (constraints): 
    (0): WordEmbeddingDistance(
        (embedding):  WordEmbedding
        (min_cos_sim):  0.5
        (cased):  False
        (include_unknown_words):  True
        (compare_against_original):  True
      )
    (1): PartOfSpeech(
        (tagger_type):  nltk
        (tagset):  universal
        (allow_verb_noun_swap):  True
        (compare_against_original):  True
      )
    (2): UniversalSentenceEncoder(
        (metric):  angular
        (threshold):  0.840845057
        (window_size):  15
        (skip_text_shorter_than_window):  True
        (compare_against_original):  False
      )
    (3): RepeatModification
    (4): StopwordModification
    (5): InputColumnModification(
        (matching_column_labels):  ['premise', 'hypothesis']
       

[Succeeded / Failed / Skipped / Total] 1 / 0 / 0 / 1:  10%|█         | 1/10 [00:00<00:08,  1.08it/s]

--------------------------------------------- Result 1 ---------------------------------------------
[[0 (98%)]] --> [[1 (69%)]]

im [[working]] [[technical]] support voice process

im [[cooperated]] [[artsy]] support voice process




[Succeeded / Failed / Skipped / Total] 2 / 0 / 0 / 2:  20%|██        | 2/10 [00:04<00:18,  2.37s/it]

--------------------------------------------- Result 2 ---------------------------------------------
[[1 (99%)]] --> [[0 (51%)]]

dirty pictureblyk on aircel thanks you for being a [[valued]] [[member]] [[heres]] an exclusive animated trailer for the dirty picture the [[insider]] scoop saysvidya balans [[sexy]] avatar itsy bitsy clothes cleavageshowing blouses and raunchy dance numbers in the dirty picture are receiving jawdropping reactions [[fans]] and [[trade]] pundits say the [[masala]] movie is expected to beat the musical treat rockstar the funfilled desi boyz and the action thriller don 2 dont miss it!

dirty pictureblyk on aircel thanks you for being a [[assessing]] [[deputies]] [[hehe]] an exclusive animated trailer for the dirty picture the [[townsend]] scoop saysvidya balans [[warmer]] avatar itsy bitsy clothes cleavageshowing blouses and raunchy dance numbers in the dirty picture are receiving jawdropping reactions [[partisans]] and [[handel]] pundits say the [[garam]] movi

[Succeeded / Failed / Skipped / Total] 3 / 0 / 0 / 3:  30%|███       | 3/10 [00:07<00:16,  2.35s/it]

--------------------------------------------- Result 3 ---------------------------------------------
[[1 (100%)]] --> [[0 (53%)]]

has your [[mortgage]] search got you down are you frustrated and confused with all the [[different]] terms and quotes don t know who is telling you the truth we can solve all your [[problems]] visit our site today and in two minutes you can have us searching [[thousands]] of programs and lenders for you get the truth get the facts get your options all in one [[shot]] it s absolutely [[free]] and you can be done in only two minutes so hyperlink click right now and put your worries behind you ryte 1635465 po1635465 kj 1635465j1635465bjk

has your [[refinance]] search got you down are you frustrated and confused with all the [[variable]] terms and quotes don t know who is telling you the truth we can solve all your [[matter]] visit our site today and in two minutes you can have us searching [[mil]] of programs and lenders for you get the truth get the facts ge

[Succeeded / Failed / Skipped / Total] 4 / 0 / 0 / 4:  40%|████      | 4/10 [00:35<00:53,  8.97s/it]

--------------------------------------------- Result 4 ---------------------------------------------
[[1 (100%)]] --> [[0 (51%)]]

[[hi]] my name is bailey and i ' [[m]] 19 years [[old]] . i [[have]] [[voluptuous]] [[body]] , [[dark]] brown [[hair]] and [[hazel]] [[eyes]] and a [[nice]] [[dd]] [[breast]] i love [[affection]] and [[looking]] for [[someone]] to [[have]] some [[fun]] with ( without [[obligations]] ) . you can [[contact]] me now at : http : / / vicinal . myabsinth . com / 575 [[r]] . html ( registration is free of [[charge]] ) [[see]] you soon you ' [[re]] a [[catch]] . you know it and we know it . now it ' s [[time]] to [[let]] them [[know]] it . 1 , 451 , 004 swingers already [[registered]] [[so]] , what are you [[waiting]] for ? just [[click]] below and [[dive]] right into [[creating]] your [[ad]] : http : / / muscular . myabsinth . com / 575 [[r]] . html if you [[got]] this [[message]] by mistake , or you do not [[wish]] to [[get]] [[messages]] from [[dating]] please [

[Succeeded / Failed / Skipped / Total] 5 / 0 / 0 / 5:  50%|█████     | 5/10 [00:54<00:54, 10.94s/it]

--------------------------------------------- Result 5 ---------------------------------------------
[[1 (100%)]] --> [[0 (63%)]]

windowflower . com is one of the [[biggest]] [[arts]] [[providers]] [[based]] in [[china]] . we [[supply]] you with high - quality , [[artistic]] , and customized [[chinese]] [[paper]] - [[cut]] [[products]] ! [[add]] [[elegance]] and cultural [[touch]] to your [[life]] and [[business]] ! windowflower . [[com]] you are [[receiving]] this email [[because]] you [[opted]] - in to [[receive]] [[special]] [[offers]] through a [[partner]] [[website]] . if you [[feel]] that you [[received]] this email in [[error]] or do not [[wish]] to [[receive]] additional [[special]] [[offers]] , please [[enter]] your [[email]] [[address]] here and [[click]] the [[button]] of 

windowflower . com is one of the [[longer]] [[artwork]] [[contributors]] [[bedrock]] in [[hua]] . we [[provides]] you with high - quality , [[eclectic]] , and customized [[shanghainese]] [[journal]] - [[

[Succeeded / Failed / Skipped / Total] 6 / 0 / 0 / 6:  60%|██████    | 6/10 [00:58<00:38,  9.73s/it]

--------------------------------------------- Result 6 ---------------------------------------------
[[0 (100%)]] --> [[1 (51%)]]

please [[join]] [[us]] for breakfast this [[friday]] , [[october]] 27 , ( by [[steve]] weller ' s [[desk]] ) around 8 : 30 a . m . to celebrate the [[following]] [[birthdays]] : [[october]] [[julie]] [[armstrong]] 15 th theresa branney 16 [[th]] don powell 20 [[th]] morgan gottsponer 22 nd t . k . lohman 29 [[th]] dana jones 31 [[st]]

please [[unites]] [[our]] for breakfast this [[sunday]] , [[avril]] 27 , ( by [[stephan]] weller ' s [[chambre]] ) around 8 : 30 a . m . to celebrate the [[next]] [[feasts]] : [[avril]] [[jolie]] [[apollo]] 15 th theresa branney 16 [[gerardo]] don powell 20 [[gonzalez]] morgan gottsponer 22 nd t . k . lohman 29 [[gerardo]] dana jones 31 [[thoroughfare]]




[Succeeded / Failed / Skipped / Total] 7 / 0 / 0 / 7:  70%|███████   | 7/10 [01:01<00:26,  8.76s/it]

--------------------------------------------- Result 7 ---------------------------------------------
[[0 (100%)]] --> [[1 (55%)]]

[[join]] clyde drexler on espeak at ethink . [[enron]] . com , wednesday , june 14 at 10 a . [[m]] . [[houston]] time . clyde , an nba legend , will conduct an " open mike " session to answer whatever questions you have for him . are you at a remote location or can ' t make the event ? go into espeak now and pre - submit your question ( s ) for clyde to answer during the [[scheduled]] [[event]] . we want to answer everyone ' s questions , but due to the high volume of questions we anticipate on this session , it would be helpful if you can keep your questions short and simple . this will increase the opportunity for your question to be answered . ethink : invest your mind

[[endorses]] clyde drexler on espeak at ethink . [[ponzi]] . com , wednesday , june 14 at 10 a . [[trillions]] . [[chandler]] time . clyde , an nba legend , will conduct an " open mike " 

[Succeeded / Failed / Skipped / Total] 8 / 0 / 0 / 8:  80%|████████  | 8/10 [01:02<00:15,  7.86s/it]

--------------------------------------------- Result 8 ---------------------------------------------
[[1 (92%)]] --> [[0 (56%)]]

[[n]] 45days [[reduce]] fat8 inchusesauna [[slim]] belt650 vibration belt999 ncr50rs extrafree [[yoko]] call now:9268250573/931130487

[[nitrogen]] 45days [[shrinkage]] fat8 inchusesauna [[fina]] belt650 vibration belt999 ncr50rs extrafree [[matsumoto]] call now:9268250573/931130487




[Succeeded / Failed / Skipped / Total] 9 / 0 / 0 / 9:  90%|█████████ | 9/10 [01:40<00:11, 11.17s/it]

--------------------------------------------- Result 9 ---------------------------------------------
[[1 (100%)]] --> [[0 (51%)]]

googlecash gives you all the tools you [[need]] to turn the search [[engine]] [[google]] . [[com]] into an autopilot [[cash]] generating machine ! what ' s your dream [[lifestyle]] ? challenge schnabel behind jolla cozy buttrick isabel [[solder]] excerpt [[breakfast]] [[tame]] firewall [[hatchet]] [[ooze]] boggle decompression [[london]] recriminate [[coat]] inbreed [[churchyard]] dar [[yoke]] [[snappy]] extrema [[grind]] [[cookie]] longitudinal [[suntan]] buttress [[herpes]] quadrillion improvisate [[idiot]] antarctica maybe [[embarrass]] bermuda lawman octillion [[musket]] betoken [[alacrity]] [[request]] [[seizure]] [[fateful]] [[mile]] conjunct [[courier]] administrable there [[stepmother]] boycott importune [[chronicle]] cavern [[catch]] aeronautic [[prestigious]] thermofax [[interruption]] irredeemable hesitater [[pursue]] abbas [[aruba]] [[berra]] de

[Succeeded / Failed / Skipped / Total] 10 / 0 / 0 / 10: 100%|██████████| 10/10 [01:42<00:00, 10.23s/it]

--------------------------------------------- Result 10 ---------------------------------------------
[[0 (100%)]] --> [[1 (58%)]]

hi does anyone [[have]] a collection of real world spam that i could use to test my [[setup]] this url email is sponsored by osdn tired of that same old cell phone get a new here for free url spamassassin talk mailing [[list]] spamassassin talk url url

hi does anyone [[enjoys]] a collection of real world spam that i could use to test my [[fixture]] this url email is sponsored by osdn tired of that same old cell phone get a new here for free url spamassassin talk mailing [[registered]] spamassassin talk url url



+-------------------------------+--------+
| Attack Results                |        |
+-------------------------------+--------+
| Number of successful attacks: | 10     |
| Number of failed attacks:     | 0      |
| Number of skipped attacks:    | 0      |
| Original accuracy:            | 100.0% |
| Accuracy under attack:        | 0.0%   |
| At

## Apply WordBug attack

In [ ]:
# WordBug attack introduces character-level perturbations

from textattack.attack_recipes import DeepWordBugGao2018
attack_2 = DeepWordBugGao2018.build(wrapped_model)

textattack: Unknown if model of class <class 'transformers.models.mobilebert.modeling_mobilebert.MobileBertForSequenceClassification'> compatible with goal function <class 'textattack.goal_functions.classification.untargeted_classification.UntargetedClassification'>.


In [ ]:
# Run the attack to the 10 sentences from the test dataset
attacker = Attacker(attack_2, testdataset10samples_dataset)
attacker.attack_dataset()

Attack(
  (search_method): GreedyWordSwapWIR(
    (wir_method):  unk
  )
  (goal_function):  UntargetedClassification
  (transformation):  CompositeTransformation(
    (0): WordSwapNeighboringCharacterSwap(
        (random_one):  True
      )
    (1): WordSwapRandomCharacterSubstitution(
        (random_one):  True
      )
    (2): WordSwapRandomCharacterDeletion(
        (random_one):  True
      )
    (3): WordSwapRandomCharacterInsertion(
        (random_one):  True
      )
    )
  (constraints): 
    (0): LevenshteinEditDistance(
        (max_edit_distance):  30
        (compare_against_original):  True
      )
    (1): RepeatModification
    (2): StopwordModification
  (is_black_box):  True
) 



[Succeeded / Failed / Skipped / Total] 1 / 0 / 0 / 1:  10%|█         | 1/10 [00:00<00:03,  2.92it/s]

--------------------------------------------- Result 1 ---------------------------------------------
[[0 (98%)]] --> [[1 (66%)]]

im [[working]] [[technical]] [[support]] voice [[process]]

im [[lworking]] [[techcical]] [[gsupport]] voice [[procpss]]




[Succeeded / Failed / Skipped / Total] 2 / 0 / 0 / 2:  20%|██        | 2/10 [00:02<00:10,  1.37s/it]

--------------------------------------------- Result 2 ---------------------------------------------
[[1 (99%)]] --> [[0 (52%)]]

dirty [[pictureblyk]] on [[aircel]] thanks you for being a [[valued]] [[member]] [[heres]] an [[exclusive]] animated trailer for the dirty picture the [[insider]] scoop saysvidya balans [[sexy]] [[avatar]] itsy bitsy clothes cleavageshowing blouses and raunchy dance numbers in the dirty picture are receiving [[jawdropping]] reactions [[fans]] and [[trade]] pundits say the [[masala]] movie is expected to beat the musical [[treat]] [[rockstar]] the funfilled desi [[boyz]] and the action thriller don 2 dont [[miss]] it!

dirty [[picturelbyk]] on [[Tircel]] thanks you for being a [[vdalued]] [[ember]] [[hers]] an [[xclusive]] animated trailer for the dirty picture the [[inlider]] scoop saysvidya balans [[sey]] [[avatTar]] itsy bitsy clothes cleavageshowing blouses and raunchy dance numbers in the dirty picture are receiving [[jaswdropping]] reactions [[fan]] and

[Succeeded / Failed / Skipped / Total] 3 / 0 / 0 / 3:  30%|███       | 3/10 [00:03<00:08,  1.24s/it]

--------------------------------------------- Result 3 ---------------------------------------------
[[1 (100%)]] --> [[0 (51%)]]

has your [[mortgage]] search got you down are you frustrated and confused with all the different terms and quotes don t know who is telling you the truth we can solve all your problems visit our site today and in two minutes you can have us searching [[thousands]] of programs and [[lenders]] for you get the truth get the facts get your options all in one shot it s absolutely free and you can be done in only two minutes so hyperlink click right now and put your worries [[behind]] you ryte 1635465 po1635465 kj 1635465j1635465bjk

has your [[mortgae]] search got you down are you frustrated and confused with all the different terms and quotes don t know who is telling you the truth we can solve all your problems visit our site today and in two minutes you can have us searching [[thousnads]] of programs and [[lnders]] for you get the truth get the facts get your

[Succeeded / Failed / Skipped / Total] 3 / 1 / 0 / 4:  40%|████      | 4/10 [00:08<00:12,  2.05s/it]

--------------------------------------------- Result 4 ---------------------------------------------
[[1 (100%)]] --> [[[FAILED]]]

hi my name is bailey and i ' m 19 years old . i have voluptuous body , dark brown hair and hazel eyes and a nice dd breast i love affection and looking for someone to have some fun with ( without obligations ) . you can contact me now at : http : / / vicinal . myabsinth . com / 575 r . html ( registration is free of charge ) see you soon you ' re a catch . you know it and we know it . now it ' s time to let them know it . 1 , 451 , 004 swingers already registered so , what are you waiting for ? just click below and dive right into creating your ad : http : / / muscular . myabsinth . com / 575 r . html if you got this message by mistake , or you do not wish to get messages from dating please click : http : / / quash . myabsinth . com / nothanks . php ion marketing limited d 2 , 23 , borrett road , mid - levels west hong kong 




[Succeeded / Failed / Skipped / Total] 3 / 2 / 0 / 5:  50%|█████     | 5/10 [00:11<00:11,  2.38s/it]

--------------------------------------------- Result 5 ---------------------------------------------
[[1 (100%)]] --> [[[FAILED]]]

windowflower . com is one of the biggest arts providers based in china . we supply you with high - quality , artistic , and customized chinese paper - cut products ! add elegance and cultural touch to your life and business ! windowflower . com you are receiving this email because you opted - in to receive special offers through a partner website . if you feel that you received this email in error or do not wish to receive additional special offers , please enter your email address here and click the button of 




[Succeeded / Failed / Skipped / Total] 4 / 2 / 0 / 6:  60%|██████    | 6/10 [00:13<00:08,  2.24s/it]

--------------------------------------------- Result 6 ---------------------------------------------
[[0 (100%)]] --> [[1 (53%)]]

[[please]] [[join]] us for breakfast this friday , [[october]] [[27]] , ( by [[steve]] [[weller]] ' s [[desk]] ) [[around]] 8 : 30 a . m . to celebrate the following [[birthdays]] : [[october]] [[julie]] [[armstrong]] 15 th theresa [[branney]] [[16]] th don powell 20 th morgan gottsponer 22 nd t . k . lohman 29 th dana jones 31 st

[[pglease]] [[jqin]] us for breakfast this friday , [[cotober]] [[72]] , ( by [[steva]] [[eller]] ' s [[deZsk]] ) [[raound]] 8 : 30 a . m . to celebrate the following [[vbirthdays]] : [[soctober]] [[ulie]] [[arfmstrong]] 15 th theresa [[braney]] [[61]] th don powell 20 th morgan gottsponer 22 nd t . k . lohman 29 th dana jones 31 st




[Succeeded / Failed / Skipped / Total] 5 / 2 / 0 / 7:  70%|███████   | 7/10 [00:15<00:06,  2.18s/it]

--------------------------------------------- Result 7 ---------------------------------------------
[[0 (100%)]] --> [[1 (53%)]]

[[join]] [[clyde]] drexler on espeak at [[ethink]] . [[enron]] . com , wednesday , [[june]] 14 at [[10]] a . m . [[houston]] time . [[clyde]] , an nba [[legend]] , will conduct an " open mike " session to answer whatever questions you have for him . are you at a remote [[location]] or can ' t make the event ? go into espeak now and pre - submit your question ( s ) for clyde to answer during the scheduled [[event]] . we want to answer everyone ' s questions , but due to the high volume of questions we anticipate on this [[session]] , it would be helpful if you can keep your questions short and simple . this will increase the opportunity for your question to be answered . [[ethink]] : invest your mind

[[Koin]] [[clytde]] drexler on espeak at [[ethnik]] . [[nron]] . com , wednesday , [[jnue]] 14 at [[1G0]] a . m . [[housotn]] time . [[cldye]] , an nba [[Llege

[Succeeded / Failed / Skipped / Total] 6 / 2 / 0 / 8:  80%|████████  | 8/10 [00:16<00:04,  2.03s/it]

--------------------------------------------- Result 8 ---------------------------------------------
[[1 (92%)]] --> [[0 (64%)]]

n [[45days]] [[reduce]] fat8 [[inchusesauna]] [[slim]] belt650 vibration [[belt999]] [[ncr50rs]] [[extrafree]] yoko [[call]] now:[[9268250573]]/[[931130487]]

n [[4p5days]] [[rEeduce]] fat8 [[incuhsesauna]] [[sim]] belt650 vibration [[belt9M99]] [[ncr05rs]] [[extraree]] yoko [[acll]] now:[[926825057]]/[[31130487]]




[Succeeded / Failed / Skipped / Total] 6 / 3 / 0 / 9:  90%|█████████ | 9/10 [00:23<00:02,  2.57s/it]

--------------------------------------------- Result 9 ---------------------------------------------
[[1 (100%)]] --> [[[FAILED]]]

googlecash gives you all the tools you need to turn the search engine google . com into an autopilot cash generating machine ! what ' s your dream lifestyle ? challenge schnabel behind jolla cozy buttrick isabel solder excerpt breakfast tame firewall hatchet ooze boggle decompression london recriminate coat inbreed churchyard dar yoke snappy extrema grind cookie longitudinal suntan buttress herpes quadrillion improvisate idiot antarctica maybe embarrass bermuda lawman octillion musket betoken alacrity request seizure fateful mile conjunct courier administrable there stepmother boycott importune chronicle cavern catch aeronautic prestigious thermofax interruption irredeemable hesitater pursue abbas aruba berra dean lockwood jimenez sioux ac collocation dempsey bowstring maintenance orono extant vast debunk galvanometer punster swatch cryptology contain conv

[Succeeded / Failed / Skipped / Total] 7 / 3 / 0 / 10: 100%|██████████| 10/10 [00:23<00:00,  2.39s/it]

--------------------------------------------- Result 10 ---------------------------------------------
[[0 (100%)]] --> [[1 (63%)]]

hi does anyone have a collection of real world spam that i could use to test my [[setup]] this [[url]] email is sponsored by osdn tired of that same old cell phone get a new here for free url [[spamassassin]] talk mailing list [[spamassassin]] talk url [[url]]

hi does anyone have a collection of real world spam that i could use to test my [[seutp]] this [[ujrl]] email is sponsored by osdn tired of that same old cell phone get a new here for free url [[spamasFsassin]] talk mailing list [[spamasstssin]] talk url [[ur]]



+-------------------------------+--------+
| Attack Results                |        |
+-------------------------------+--------+
| Number of successful attacks: | 7      |
| Number of failed attacks:     | 3      |
| Number of skipped attacks:    | 0      |
| Original accuracy:            | 100.0% |
| Accuracy under attack:        | 30.0% 

## Apply PWWS (Probability Weighted Word Saliency) attack

In [ ]:
# PWWS attack uses word importance scores to modify text (i.e., modify the most important words in a sentence)
from textattack.attack_recipes import PWWSRen2019
attack_3 = PWWSRen2019.build(wrapped_model)

[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
textattack: Unknown if model of class <class 'transformers.models.mobilebert.modeling_mobilebert.MobileBertForSequenceClassification'> compatible with goal function <class 'textattack.goal_functions.classification.untargeted_classification.UntargetedClassification'>.


In [ ]:
# Run the attack to the 10 sentences from the test dataset
attacker = Attacker(attack_3, testdataset10samples_dataset)
attacker.attack_dataset()

Attack(
  (search_method): GreedyWordSwapWIR(
    (wir_method):  weighted-saliency
  )
  (goal_function):  UntargetedClassification
  (transformation):  WordSwapWordNet
  (constraints): 
    (0): RepeatModification
    (1): StopwordModification
  (is_black_box):  True
) 



[Succeeded / Failed / Skipped / Total] 1 / 0 / 0 / 1:  10%|█         | 1/10 [00:01<00:14,  1.59s/it]

--------------------------------------------- Result 1 ---------------------------------------------
[[0 (98%)]] --> [[1 (98%)]]

im working technical [[support]] voice process

im working technical [[patronize]] voice process




[Succeeded / Failed / Skipped / Total] 1 / 1 / 0 / 2:  20%|██        | 2/10 [00:13<00:52,  6.59s/it]

--------------------------------------------- Result 2 ---------------------------------------------
[[1 (99%)]] --> [[[FAILED]]]

dirty pictureblyk on aircel thanks you for being a valued member heres an exclusive animated trailer for the dirty picture the insider scoop saysvidya balans sexy avatar itsy bitsy clothes cleavageshowing blouses and raunchy dance numbers in the dirty picture are receiving jawdropping reactions fans and trade pundits say the masala movie is expected to beat the musical treat rockstar the funfilled desi boyz and the action thriller don 2 dont miss it!




[Succeeded / Failed / Skipped / Total] 2 / 1 / 0 / 3:  30%|███       | 3/10 [00:27<01:03,  9.12s/it]

--------------------------------------------- Result 3 ---------------------------------------------
[[1 (100%)]] --> [[0 (55%)]]

has your mortgage search [[got]] you down are you frustrated and confused with all the different terms and quotes don t know who is telling you the truth we can solve all your problems [[visit]] our site today and in two minutes you can have us searching [[thousands]] of [[programs]] and lenders for you [[get]] the truth get the facts [[get]] your options all in one [[shot]] it s absolutely [[free]] and you can be done in only two minutes so hyperlink click right now and [[put]] your worries [[behind]] you ryte 1635465 po1635465 kj 1635465j1635465bjk

has your mortgage search [[scram]] you down are you frustrated and confused with all the different terms and quotes don t know who is telling you the truth we can solve all your problems [[chatter]] our site today and in two minutes you can have us searching [[K]] of [[plan]] and lenders for you [[start]] the 

[Succeeded / Failed / Skipped / Total] 2 / 2 / 0 / 4:  40%|████      | 4/10 [00:51<01:16, 12.76s/it]

--------------------------------------------- Result 4 ---------------------------------------------
[[1 (100%)]] --> [[[FAILED]]]

hi my name is bailey and i ' m 19 years old . i have voluptuous body , dark brown hair and hazel eyes and a nice dd breast i love affection and looking for someone to have some fun with ( without obligations ) . you can contact me now at : http : / / vicinal . myabsinth . com / 575 r . html ( registration is free of charge ) see you soon you ' re a catch . you know it and we know it . now it ' s time to let them know it . 1 , 451 , 004 swingers already registered so , what are you waiting for ? just click below and dive right into creating your ad : http : / / muscular . myabsinth . com / 575 r . html if you got this message by mistake , or you do not wish to get messages from dating please click : http : / / quash . myabsinth . com / nothanks . php ion marketing limited d 2 , 23 , borrett road , mid - levels west hong kong 




[Succeeded / Failed / Skipped / Total] 2 / 3 / 0 / 5:  50%|█████     | 5/10 [01:03<01:03, 12.69s/it]

--------------------------------------------- Result 5 ---------------------------------------------
[[1 (100%)]] --> [[[FAILED]]]

windowflower . com is one of the biggest arts providers based in china . we supply you with high - quality , artistic , and customized chinese paper - cut products ! add elegance and cultural touch to your life and business ! windowflower . com you are receiving this email because you opted - in to receive special offers through a partner website . if you feel that you received this email in error or do not wish to receive additional special offers , please enter your email address here and click the button of 




[Succeeded / Failed / Skipped / Total] 3 / 3 / 0 / 6:  60%|██████    | 6/10 [01:07<00:45, 11.27s/it]

--------------------------------------------- Result 6 ---------------------------------------------
[[0 (100%)]] --> [[1 (54%)]]

[[please]] [[join]] [[us]] for breakfast this friday , october [[27]] , ( by steve [[weller]] ' s desk ) [[around]] [[8]] : 30 a . m . to [[celebrate]] the [[following]] birthdays : [[october]] julie armstrong [[15]] th theresa branney [[16]] th don powell [[20]] [[th]] morgan gottsponer [[22]] [[nd]] t . k . lohman 29 th dana jones [[31]] st

[[delight]] [[articulation]] [[uracil]] for breakfast this friday , october [[twenty-seven]] , ( by steve [[good]] ' s desk ) [[roughly]] [[octonary]] : 30 a . m . to [[fete]] the [[chase]] birthdays : [[Oct]] julie armstrong [[XV]] th theresa branney [[sixteen]] th don powell [[twenty]] [[thorium]] morgan gottsponer [[XXII]] [[neodymium]] t . k . lohman 29 th dana jones [[thirty-one]] st




[Succeeded / Failed / Skipped / Total] 3 / 4 / 0 / 7:  70%|███████   | 7/10 [01:23<00:35, 11.86s/it]

--------------------------------------------- Result 7 ---------------------------------------------
[[0 (100%)]] --> [[[FAILED]]]

join clyde drexler on espeak at ethink . enron . com , wednesday , june 14 at 10 a . m . houston time . clyde , an nba legend , will conduct an " open mike " session to answer whatever questions you have for him . are you at a remote location or can ' t make the event ? go into espeak now and pre - submit your question ( s ) for clyde to answer during the scheduled event . we want to answer everyone ' s questions , but due to the high volume of questions we anticipate on this session , it would be helpful if you can keep your questions short and simple . this will increase the opportunity for your question to be answered . ethink : invest your mind




[Succeeded / Failed / Skipped / Total] 4 / 4 / 0 / 8:  80%|████████  | 8/10 [01:24<00:21, 10.59s/it]

--------------------------------------------- Result 8 ---------------------------------------------
[[1 (92%)]] --> [[0 (52%)]]

[[n]] 45days reduce fat8 inchusesauna slim belt650 [[vibration]] belt999 ncr50rs extrafree yoko [[call]] now:9268250573/931130487

[[nitrogen]] 45days reduce fat8 inchusesauna slim belt650 [[oscillation]] belt999 ncr50rs extrafree yoko [[squall]] now:9268250573/931130487




[Succeeded / Failed / Skipped / Total] 4 / 5 / 0 / 9:  90%|█████████ | 9/10 [01:48<00:12, 12.08s/it]

--------------------------------------------- Result 9 ---------------------------------------------
[[1 (100%)]] --> [[[FAILED]]]

googlecash gives you all the tools you need to turn the search engine google . com into an autopilot cash generating machine ! what ' s your dream lifestyle ? challenge schnabel behind jolla cozy buttrick isabel solder excerpt breakfast tame firewall hatchet ooze boggle decompression london recriminate coat inbreed churchyard dar yoke snappy extrema grind cookie longitudinal suntan buttress herpes quadrillion improvisate idiot antarctica maybe embarrass bermuda lawman octillion musket betoken alacrity request seizure fateful mile conjunct courier administrable there stepmother boycott importune chronicle cavern catch aeronautic prestigious thermofax interruption irredeemable hesitater pursue abbas aruba berra dean lockwood jimenez sioux ac collocation dempsey bowstring maintenance orono extant vast debunk galvanometer punster swatch cryptology contain conv

[Succeeded / Failed / Skipped / Total] 5 / 5 / 0 / 10: 100%|██████████| 10/10 [01:53<00:00, 11.37s/it]

--------------------------------------------- Result 10 ---------------------------------------------
[[0 (100%)]] --> [[1 (78%)]]

hi does anyone have a collection of real world spam that i could [[use]] to test my setup this url email is [[sponsored]] by osdn tired of that same [[old]] cell phone get a new here for free url spamassassin talk mailing [[list]] spamassassin talk url url

hi does anyone have a collection of real world spam that i could [[habituate]] to test my setup this url email is [[patronise]] by osdn tired of that same [[honest-to-goodness]] cell phone get a new here for free url spamassassin talk mailing [[heel]] spamassassin talk url url



+-------------------------------+--------+
| Attack Results                |        |
+-------------------------------+--------+
| Number of successful attacks: | 5      |
| Number of failed attacks:     | 5      |
| Number of skipped attacks:    | 0      |
| Original accuracy:            | 100.0% |
| Accuracy under attack:    

## Adversarial Attack Report

**TextFooler:** This adversarial attack identifies critical words in a text using importance score and replaces them with synonyms from embedding spaces. It prioritizes substitutions that preserve sentence fluency and meaning while causing misclassifications. The goal is deceive the model with minimal, human imperceptible perturbations.

**WordBug:** This method generates adversarial attacks by applying character-level perturbations to words most impactful on model confidence. It ranks words by their influence on prediction certainty then introduces subtle edits to mislead the model. The attack balances stealth and effectiveness by limiting visible distortions.

**PWWS:** This attack combines the impact on predictions and and class probabilities to select substitution targets, prioritzing high impact words. It replaces these with WordNet synonyms, weighted by their likelihood to alter model output while ensuring semantic coherence. The approach systematically maximizes misslcassification by optimizing both word importance and substitution efficacy.


**Attack Analysis**

The TextFooler attack had a 100% success rate, however it was easy to tell that something was off with the texts. Changing "im [[working]] [[technical]] support voice process" to "im [[cooperated]] [[artsy]] support voice process" will change how the model reads them but also changes a message that is "ham" to something that reads as "spam". If someone was anaylising the messages and saw that this was flagged as incorrect they would quickly check to see if the data being plugged in is usable and see the attack.

The WordBug attack had a 70% success rate. Although each individual change would be noticeable if isolated, people have a tendency to overlook typos when reading sentences. Therefore the likelyhood of someone noticing the change from "working" to "lworking" is very unlikely in my opinion.

The PWWS attack had a 50% success rate, however it altered substantially less words and much less noticeably than the other attacks. Sticking to synonyms that maintain semantic coherence guarantees that the result is hard to distiguish from an actual message.